openrarity/collection/collection.py

```diff
@@ -0,0 +1,337 @@
+import json
+from hashlib import md5
+from itertools import chain
+from logging import Logger
+from pathlib import Path
+from typing import Literal, TypedDict, cast, overload
+
+from satchel.aggregate import groupapply
+
+from openrarity.io import read, write
+from openrarity.metrics.ic import calculate_entropy, information_content
+from openrarity.token import (
+    AttributeStatistic,
+    RankedToken,
+    RawToken,
+    TokenId,
+    TokenStatistic,
+    ValidatedTokenAttribute,
+    enforce_schema,
+    validate_tokens,
+)
+from openrarity.token.metadata.dtypes import (
+    process_numeric_dtypes,
+    process_string_dtypes,
+)
+from openrarity.types import JsonEncodable
+from openrarity.utils import merge, rank_over
+
+from .utils import aggregate_tokens, flatten_token_data
+
+logger = Logger(__name__)
+
+
+class AttributesStatistics(TypedDict):
+    string: list[AttributeStatistic]
+    number: list[AttributeStatistic]
+    date: list[AttributeStatistic]
+
+
+class TokenCollection:
+    """ """
+
+    def __init__(
+        self,
+        token_type: Literal["non-fungible", "semi-fungible"],
+        tokens: dict[TokenId, RawToken],
+    ):
+        self._token_type = token_type
+        self._input_checksum: str = self._hash_data(cast(JsonEncodable, tokens))
+        self._ranks_checksum: str | None = None
+        self._token_supply: int | dict[str | int, int]
+        (self._token_supply, self._tokens) = validate_tokens(token_type, tokens)
+
+        # Derived data
+        self._vertical_attribute_data: list[ValidatedTokenAttribute] = []
+        self._attribute_statistics: AttributesStatistics = {
+            "string": [],
+            "number": [],
+            "date": [],
+        }
+        self._entropy: float | None = None
+        self._token_statistics: list[TokenStatistic] = []
+        self._string_types: list[ValidatedTokenAttribute] = []
+        self._number_types: list[ValidatedTokenAttribute] = []
+        self._date_types: list[ValidatedTokenAttribute] = []
+
+        # Output data
+        self._ranks: list[RankedToken] = []
+
+    def __repr__(self) -> str:
+        return f"Collection({self._token_type})"
+
+    @property
+    def tokens(self) -> dict[TokenId, RawToken]:
+        return self._tokens
+
+    @property
+    def total_supply(self) -> int:
+        # SemiFungible needs to sum the supply of each token
+        if isinstance(self._token_supply, dict):
+            return sum(self._token_supply.values())
+
+        return self._token_supply
+
+    @property
+    def attribute_statistics(self) -> AttributesStatistics:
+        if not self._attribute_statistics:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._attribute_statistics
+
+    @property
+    def entropy(self) -> float | None:
+        return self._entropy
```

Comment:
entropy is also computed and should follow the same assignment pattern in the code, no?

Reply:
In this `self._entropy` is the calculated value and using `@property` to decorate a method like this effectively creates a read only variable.